### ISSUE TITLE
Literal assignment in return position ignores type declaration

### ISSUE THREAD

**Issue by s-broda** [JuliaLang/julia#18948]
Should this be an error?

```
julia> foo()=(local bar::Int64; bar=1.5)
foo (generic function with 1 method)

julia> foo()
1.5
```

versioninfo:
```
Julia Version 0.5.0
Commit 3c9d753 (2016-09-19 18:14 UTC)
Platform Info:
  System: NT (x86_64-w64-mingw32)
  CPU: Intel(R) Processor 5Y10 CPU @ 0.80GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.7.1 (ORCJIT, broadwell)
```

**Comment by yuyichao** [comment]
Fixed on master.

**Comment by yuyichao** [comment]
Errr, the fix isn't really backportable.....

The change in checked convertion makes the throw explicit to type inference so that it doesn't infer `checked_fptosi` as pure anymore. We should probably check if any of the other intrinsics have this problem too.

**Comment by JeffBezanson** [comment]
This is a serious bug. Needs to be fixed on the 0.5.x branch somehow.

**Comment by tkelman** [comment]
backport label redundant here since it's on the pr #18955
